src/fidesops/graph/graph_differences.py

```diff
@@ -0,0 +1,200 @@
+from typing import Any, Dict, List, Optional, Set
+
+from fidesops.graph.config import (
+    ROOT_COLLECTION_ADDRESS,
+    TERMINATOR_ADDRESS,
+    CollectionAddress,
+)
+from fidesops.schemas.base_class import BaseSchema
+from fidesops.util.collection_util import Row
+
+GraphRepr = Dict[str, Dict[str, List[str]]]
+
+
+def format_graph_for_caching(
+    env: Dict[CollectionAddress, Any], end_nodes: List[CollectionAddress]
+) -> GraphRepr:
+    """
+    Builds a representation of the current graph (that includes its edges) for caching in Redis.
+
+    Requires the results of traversal.traverse():
+        - the modified `env`
+        - and the outputted end_nodes, which are the final nodes without children
+
+    Maps collections to their upstream dependencies and associated edges. The root is stored as having
+    no upstream collections and the terminator collection has no incoming edges.
+
+    Example:
+    {
+       <collection>: {
+          <upstream collection>: [edge between upstream and current],
+          <another upstream collection>: [edge between another upstream and current]
+       },
+       <root collection>: {},
+       <terminator collection>: {
+            <end collection>: [],
+            <end collection>: []
+       }
+    }
+    """
+    graph_repr: GraphRepr = {
+        collection.value: {
+            upstream_collection_address.value: [str(edge) for edge in edge_list]
+            for upstream_collection_address, edge_list in g_task.incoming_edges_by_collection.items()
+        }
+        for collection, g_task in env.items()
```

Comment:
good callout, I do actually know the specific type of `env` in this case, so I will improve.   This `env` variable can have different types in different contexts but I know what it is here.  

We've previously talked about a `dsk` variable of type:  Dict[CollectionAddress, Tuple[Any, ...]], which was tricky because the first item in the Tuple was typically a GraphTask, followed by any number of CollectionAddresses, so it was a Tuple of mixed types.  The `env` variable floating around is used to create the `dsk` variable, but that is not relevant here.